pika/adapters/utils/nbio_interface.py

```diff
@@ -0,0 +1,429 @@
+        timeout.
+
+        If two are scheduled for the same time, it's undefined which one will
+        be called first.
+
+        :param float delay: The number of seconds to wait to call callback
+        :param method callback: The callback method
+        :returns: A handle that can be used to cancel the request.
+        :rtype: AbstractTimerReference
+
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def getaddrinfo(self, host, port, on_done, family=0, socktype=0, proto=0,
+                    flags=0):
+        """Perform the equivalent of `socket.getaddrinfo()` asynchronously.
+
+        See `socket.getaddrinfo()` for the standard args.
+
+        :param callable on_done: user callback that takes the return value of
+            `socket.getaddrinfo()` upon successful completion or exception upon
+            failure (check for `BaseException`) as its only arg. It will not be
+            called if the operation was cancelled.
+        :rtype: AbstractIOReference
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def connect_socket(self, sock, resolved_addr, on_done):
+        """Perform the equivalent of `socket.connect()` on a previously-resolved
+        address asynchronously.
+
+        IMPLEMENTATION NOTE: Pika's connection logic resolves the addresses
+            prior to making socket connections, so we don't need to burden the
+            implementations of this method with the extra logic of asynchronous
+            DNS resolution. Implementations can use `socket.inet_pton()` to
+            verify the address.
+
+        :param socket.socket sock: non-blocking socket that needs to be
+            connected via `socket.socket.connect()`
+        :param tuple resolved_addr: resolved destination address/port two-tuple
+            as per `socket.socket.connect()`, except that the first element must
+            be an actual IP address that's consistent with the given socket's
+            address family.
+        :param callable on_done: user callback that takes None upon successful
+            completion or exception (check for `BaseException`) upon error as
+            its only arg. It will not be called if the operation was cancelled.
+
+        :rtype: AbstractIOReference
+        :raises ValueError: if host portion of `resolved_addr` is not an IP
+            address or is inconsistent with the socket's address family as
+            validated via `socket.inet_pton()`
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def create_streaming_connection(self,
+                                    protocol_factory,
+                                    sock,
+                                    on_done,
+                                    ssl_context=None,
+                                    server_hostname=None):
+        """Perform SSL session establishment, if requested, on the already-
```

Comment:
I was surprised that a method called `create_streaming_connection` is described with "Perform SSL session establishment"

Reply:
@lukebakken, as mentioned at the top of the script in its docstring, I modeled the interface on asyncio. `create_streaming_connection()` is a subset of the behavior of [asyncio's AbstractEventLoop.create_connection()](https://docs.python.org/3/library/asyncio-eventloop.html#creating-connections). The rationale is that a protocol (AMQP connection in our case) needs a stream. The stream might be TCP or SSL/TLS on top of TCP. The protocol doesn't care which one. It's up to the transport to abstract away the differences. Pika's new `AbstractIOServices.create_streaming_connection()` takes care of the details, just as the comparable method in `asyncio`.